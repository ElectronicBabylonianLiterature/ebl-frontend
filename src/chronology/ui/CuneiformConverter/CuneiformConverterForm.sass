textarea.form-control.assurbanipal
    font-size: 150%
    font-family: Assurbanipal, Junicode, serif  !important
textarea.form-control.neo-babylonian
    font-size: 150%
    font-family: Neo-Babylonian,'Adobe Blank', Junicode, serif  !important
