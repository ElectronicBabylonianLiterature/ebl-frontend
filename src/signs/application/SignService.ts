import Bluebird from 'bluebird'
import Sign, { SignQuery } from 'signs/domain/Sign'
import SignRepository from 'signs/infrastructure/SignRepository'
import { AnnotationToken } from 'fragmentarium/domain/annotation-token'

import { CroppedAnnotation } from 'signs/domain/CroppedAnnotation'

export default class SignService {
  private readonly signsRepository: SignRepository

  constructor(signsRepository: SignRepository) {
    this.signsRepository = signsRepository
  }

  getImages(signName: string): Bluebird<CroppedAnnotation[]> {
    return this.signsRepository.getImages(signName)
  }

  associateSigns(
    tokens: ReadonlyArray<ReadonlyArray<AnnotationToken>>
  ): Bluebird<ReadonlyArray<ReadonlyArray<AnnotationToken>>> {
    return this.signsRepository.associateSigns(tokens)
  }

  search(signQuery: SignQuery): Bluebird<Sign[]> {
    return this.signsRepository.search(signQuery)
  }

  find(signName: string): Bluebird<Sign> {
    return this.signsRepository.find(signName)
  }
}
