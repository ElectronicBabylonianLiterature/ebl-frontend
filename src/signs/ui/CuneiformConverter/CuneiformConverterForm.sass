.cuneiform-converter-form 
    textarea.form-control 
        font-size: 150%

        &.assurbanipal 
            font-family: Assurbanipal, Junicode, serif;
        

        &.esagil 
            font-family: 'Neo-Babylonian', Junicode, serif;
        

        &.santakku 
            font-family: 'Old-Babylonian Cursive', Junicode, serif;
        

        &.obfreie 
            font-family: 'Old-Babylonian Literature', Junicode, serif;
        

        &.santakkum 
            font-family: 'Old-Babylonian (Monumental)', Junicode, serif;
        

        &.ullikummia 
            font-family: 'Hittite', Junicode, serif;
        
    
