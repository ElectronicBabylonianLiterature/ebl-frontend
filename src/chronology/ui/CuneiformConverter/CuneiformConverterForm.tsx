import React, { useState } from 'react'
import { Form, Button } from 'react-bootstrap'
import Bluebird from 'bluebird'
import SignService from 'signs/application/SignService'
import replaceTransliteration from 'fragmentarium/domain/replaceTransliteration'
import { displayUnicode } from 'signs/ui/search/SignsSearch'
import './CuneiformConverterForm.sass'

function CuneiformConverterForm({
  signService,
}: {
  signService: SignService
}): JSX.Element {
  const [content, setContent] = useState('')
  const [convertedContent, setConvertedContent] = useState('')
  const [selectedFont, setSelectedFont] = useState('Assurbanipal')

  const handleChange = (event) => {
    setContent(event.target.value)
  }

  const handleConvert = () => {
    const lines = content.split('\n')
    const replacedLines = lines.map((line) => replaceTransliteration(line))

    Promise.all(
      replacedLines
        .filter((line) => line.trim() !== '')
        .map((line) => query(line))
    )
      .then((results) => {
        const convertedText = results
          .map((result) =>
            result
              .map((entry) =>
                entry.unicode[0] === 9999 ? ' ' : displayUnicode(entry.unicode)
              )
              .join('')
          )
          .join('\n')

        setConvertedContent(convertedText)
      })
      .catch((error) => {
        console.error('Query Error:', error)
      })
  }

  const query = (content) => {
    return Bluebird.resolve(signService.getUnicodeFromAtf(content))
  }

  const handleKeyDown = (event) => {
    if (event.key === 'Enter' && event.shiftKey) {
      event.preventDefault()
      handleConvert()
    }
  }

  const handleFontChange = (event) => {
    setSelectedFont(event.target.value)
  }

  return (
    <div className="cuneiform-converter-form">
      <Form.Label htmlFor="inputText">
        This tool allows to convert transliterations to Unicode cuneiform
        (ranges U+12000-U+123FF, U+12400-U+1247F, and U+12480-U+1254F), using
        the mapping from the eBL sign list. Different fonts, developed by S.
        Vanseveren, can be used to display the cuneiform text.
      </Form.Label>
      <Form.Control
        as="textarea"
        id="inputText"
        aria-label="input-atf"
        aria-describedby="textHelpBlock"
        value={content}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
      />
      <Form.Text id="textHelpBlock" muted>
        Enter the text you want to convert to Unicode.
      </Form.Text>
      <Form.Label htmlFor="fontSelector" style={{ paddingRight: '8px' }}>
        Select Font
      </Form.Label>
      <select
        id="fontSelector"
        className="form-select"
        value={selectedFont}
        onChange={handleFontChange}
      >
        <option value="Assurbanipal">Neo-Assyrian</option>
        <option value="Esagil">Neo-Babylonian</option>
        <option value="Santakku">Old Babylonian</option>
        <option value="SantakkuM">Old Babylonian Monumental</option>
        <option value="UllikummiA">Hittite</option>
      </select>
      <br></br>
      <Button onClick={handleConvert}>Convert</Button>
      <br></br>
      <Form.Label htmlFor="outputText">Converted Text</Form.Label>
      <Form.Control
        as="textarea"
        id="outputText"
        className={`${selectedFont.toLowerCase().replace(/\s/g, '-')}`}
        aria-describedby="outputHelpBlock"
        value={convertedContent}
        readOnly
      />
      <Form.Text id="outputHelpBlock" muted>
        This is the converted Unicode text.
      </Form.Text>
    </div>
  )
}

export default CuneiformConverterForm
