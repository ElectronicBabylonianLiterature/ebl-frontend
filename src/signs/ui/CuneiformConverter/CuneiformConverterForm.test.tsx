import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import CuneiformConverterForm from 'signs/ui/CuneiformConverter/CuneiformConverterForm'
import SignService from 'signs/application/SignService'

jest.mock('signs/application/SignService')
const signServiceMock = new (SignService as jest.Mock<
  jest.Mocked<SignService>
>)()

let container: HTMLElement

describe('CuneiformConverterForm', () => {
  beforeEach(() => {
    Object.defineProperty(window.navigator, 'clipboard', {
      value: {
        writeText: jest.fn().mockResolvedValue(true),
      },
      writable: true,
    })
    container = render(<CuneiformConverterForm signService={signServiceMock} />)
      .container
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('renders form, options, and scenario panel correctly', () => {
    expect(container).toMatchSnapshot()
  })
  it('handles input change', () => {
    const inputTextArea = screen.getByLabelText('input-atf')
    fireEvent.change(inputTextArea, { target: { value: 'test text' } })
    expect(inputTextArea).toHaveValue('test text')
  })
  it('handles font change', () => {
    const fontSelector = screen.getByLabelText('Select Font')
    fireEvent.change(fontSelector, { target: { value: 'Esagil' } })
    expect(fontSelector).toHaveValue('Esagil')
  })
  it('converts text correctly', async () => {
    signServiceMock.getUnicodeFromAtf.mockResolvedValueOnce([
      { unicode: [73979] },
    ])

    const inputTextArea = screen.getByLabelText('input-atf')
    fireEvent.change(inputTextArea, { target: { value: 'test text' } })

    const convertButton = screen.getByText('Convert')
    fireEvent.click(convertButton)

    await waitFor(() => {
      expect(screen.getByLabelText('Converted Text')).toHaveValue('ð’ƒ»')
    })
  })

  it('copies converted text to clipboard', async () => {
    signServiceMock.getUnicodeFromAtf.mockResolvedValueOnce([
      { unicode: [73979] },
    ])

    const inputTextArea = screen.getByLabelText('input-atf')
    fireEvent.change(inputTextArea, { target: { value: 'test text' } })

    const convertButton = screen.getByText('Convert')
    fireEvent.click(convertButton)

    await waitFor(() => {
      expect(screen.getByLabelText('Converted Text')).toHaveValue('ð’ƒ»')
    })

    const copyButton = screen.getByText('Copy')
    fireEvent.click(copyButton)

    await waitFor(() => {
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith('ð’ƒ»')
    })
  })
})
