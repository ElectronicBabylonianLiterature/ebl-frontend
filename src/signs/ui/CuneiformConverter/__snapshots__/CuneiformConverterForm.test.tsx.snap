// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CuneiformConverterForm renders form, options, and scenario panel correctly 1`] = `
<div>
  <div
    class="cuneiform-converter-form"
  >
    <label
      class="form-label"
      for="inputText"
    >
      This tool allows to convert transliterations to Unicode cuneiform (ranges U+12000-U+123FF, U+12400-U+1247F, and U+12480-U+1254F), using the mapping from the eBL sign list. Different fonts, developed by S. Vanseveren, can be used to display the cuneiform text.
    </label>
    <textarea
      aria-describedby="textHelpBlock"
      aria-label="input-atf"
      class="form-control"
      id="inputText"
    />
    <small
      class="form-text text-muted"
      id="textHelpBlock"
    >
      Enter the text you want to convert to Unicode.
    </small>
    <label
      class="form-label"
      for="fontSelector"
      style="padding-right: 8px;"
    >
      Select Font
    </label>
    <select
      class="form-select"
      id="fontSelector"
    >
      <option
        value="Assurbanipal"
      >
        Neo-Assyrian
      </option>
      <option
        value="Esagil"
      >
        Neo-Babylonian
      </option>
      <option
        value="Santakku"
      >
        Old Babylonian
      </option>
      <option
        value="SantakkuM"
      >
        Old Babylonian Monumental
      </option>
      <option
        value="UllikummiA"
      >
        Hittite
      </option>
    </select>
    <br />
    <button
      class="btn btn-primary"
      type="button"
    >
      Convert
    </button>
    <br />
    <label
      class="form-label"
      for="outputText"
    >
      Converted Text
    </label>
    <textarea
      aria-describedby="outputHelpBlock"
      class="assurbanipal form-control"
      id="outputText"
      readonly=""
    />
    <button
      class="mt-2 btn btn-primary"
      type="button"
    >
      Copy
    </button>
  </div>
</div>
`;
